### ISSUE TITLE
RFC 3.x: Better before filters

### ISSUE THREAD

**Issue by sukihub** [cakephp/cakephp#8105]
One thing that I'm missing in CakePHP is an option to use multiple before filters in my controllers. I often want to do (a) multiple unrelated operations in a before filter and (b) not have all of the operations execute before all of the actions. My code now looks like this:

```php
public function beforeFilter()
{
    if (!$this->user->is('reviewer')) {
        throw new ForbiddenException();
    }
    
    $action = $this->request->action;
    if ($action == 'edit' || $action == 'save') {
        if (!$this->user->is('editor')) {
            throw new ForbiddenException();
        }

        if ($this->getArticle()->isPublished()) {
            return $this->redirect(['controller' => 'Revisions', 'action' => 'add']);
        }
    }

    // ...
}
```
Hard to understand and very easy to break, if you ask me. 

Ruby on Rails solved this years ago with very elegant solution:
  * https://robots.thoughtbot.com/before-filter-wisdom
  * http://guides.rubyonrails.org/action_controller_overview.html#filters

Basically, instead of the code above, I imagine writing something like this:
```php
public function initialize()
{
    parent::initialize();

    $this->addBeforeFilter('denyAccess', ['unless' => 'isReviewer']);
    $this->addBeforeFilter('denyAccess', ['unless' => 'isEditor', 'only' => ['edit', 'save']]);
    $this->addBeforeFilter('newRevision', ['if' => 'isPublished', 'only' => ['edit', 'save']]);
}

protected function isReviewer()
{
    return $this->user->is('reviewer');
}

protected function isEditor()
{
    return $this->user->is('editor');
}

protected function isPublished()
{
    return $this->getArticle()->isPublished();
}

protected function denyAccess()
{
    throw new ForbiddenException();
}

protected function newRevision()
{
    return $this->redirect(['controller' => 'Revisions', 'action' => 'add']);
}
```
Before filters would be executed in the order in which they were added. Similar to Rails, if any of them redirects or renders something (can we intercept that?) (or throws an exception, obviously), the controller action would not get executed. I can also imagine an option to use anonymous functions instead of controller methods.

Backwards compatibility would be pretty easy, basically just doing this somewhere in controller initialization:
```php
$this->addBeforeFilter('beforeFilter');
```

`Controller.startup` event would execute a new method, which would iterate before filters, check their conditions and execute them. 

And, of course, after filters could be implemented in the exact same way.

If you like the idea, I can start implementing it. But first I wanted to validate the idea, hear your opinions, API suggestions, etc.

**Comment by markstory** [comment]
> if any of them redirects or renders something (can we intercept that?) 

We cannot currently catch redirects unless the response object is returned. The same would hold true for render() calls.

> Backwards compatibility would be pretty easy, basically just doing this somewhere in controller initialization:

In order to be backwards compatible. No changes should be required by developers upgrading their applications. We'd have to put any additional setup methods into the framework code.

One question I would have around this is why  `Controller.beforeFilter` is so special? Seems like improving only this method creates a unique situation and breaks the API consistency we have right now. Instead of creating special scenarios, could we look at extending the Event subsystem? In doing that we allow these kinds of improvements to be used anywhere an event listener could be used. This would allow conditional event listeners in models, or views as well. A possible but verbose API could be:

```php
$this->eventManager()->on(
  'Controller.startup',
  new ConditionalListener($callable, $this,  ['if' => ..., 'unless' => ...])
);
```

This API is a bit wordy, but could be wrapped with another method like

```php
$this->eventManager()->conditionalOn('Controller.startup', $callable, $conditions);
```

**Comment by ADmad** [comment]
Based on your example code it seems you want to use beforeFilter event to handle authorization. We already have an elegant solution for that: authorization adapters. So why not use that?

**Comment by sukihub** [comment]
> We already have an elegant solution for that: authorization adapters. So why not use that? 

That was just the simplest example that came to my mind. There are plenty of other uses:
* I use before filter to load current user in each request (and set locale and timezone)
* another example would be my base CronController where I use before filter to (1) turn off rendering and (2) insert a row into a table that tracks running processes,
* sometimes I do input input validation/sanitization there; especially in cases where redirect might happen (otherwise - in action - you either have to remember to `return` or have the rest of the code wrapped in `if`)
* and sometimes I even use it to load entity and check something (that again might end with redirect) - like the third part of my example above (redirect to new revision form if acrticle is published)

> Instead of creating special scenarios, could we look at extending the Event subsystem? In doing that we allow these kinds of improvements to be used anywhere an event listener could be used.

Sounds great :-) Even `only` and `except` conditions for actions are just special cases of `if`.

> This API is a bit wordy, but could be wrapped with another method like

I can still imagine simplifying it even more, to something like `addBeforeFilter` in my example. Rails is now starting to call these `before_action` and `after_action`, I don't know why they changed it.

> In order to be backwards compatible. No changes should be required by developers upgrading their applications. We'd have to put any additional setup methods into the framework code.

That is what I ment, I should have been more clear :-)

> We cannot currently catch redirects unless the response object is returned. The same would hold true for render() calls.

I think that returning response object in any of the before filters will do the trick.

**Comment by ADmad** [comment]
The conditional triggering of event listeners proposed by @markstory sounds good.

**Comment by jeffblack360** [comment]
I concur. Leveraging the ``Event`` sub-system and building that out where necessary offers flexibility, power, and probably zero risk to existing stakeholders.

**Comment by bravo-kernel** [comment]
:tractor:

**Comment by sukihub** [comment]
So, "the core" - `ConditionalCallable` class is ready here: https://github.com/cakephp/cakephp/compare/master...sukihub:3.x-conditional-listeners

Instead of `conditionalOn()` I use options to find out whether conditional listener should be attached:
```php
$this->eventManager()->on('Controller.startup', ['if' => ...], $callable);
```
I would also like to add some helper methods to `Controller` class, stay tuned :-)

**Comment by markstory** [comment]
@sukihub Looking good. The `getConditions` method should be able to leverage `is_callable` instead of the sniffs you're using right now.

Should the if/unless conditions get any arguments? Perhaps the could get the event as an argument?

**Comment by sukihub** [comment]
@markstory thanks, very good points :-) fixed here https://github.com/sukihub/cakephp/commit/4ac6a375c04ef4b56cf9e9b7744456aed94b12d7

And simplified before/after filter controller API is here https://github.com/sukihub/cakephp/commit/4521a943ba78624cfd72c60e692c9af2bba7afff. Not tested yet though. I went for (Rails inspired) before action / after action names, but I have no problem sticking with before filter / after filter naming.

**Comment by markstory** [comment]
Looking good @sukihub  I don't think if/unless need any arguments. The callables can be defined with the correct scope imo.

**Comment by sukihub** [comment]
Then I probably misunderstood you when you said:
> Should the if/unless conditions get any arguments? Perhaps they could get the event as an argument?

The only way to get event as an argument is to pass all the arguments of event handler.

**Comment by markstory** [comment]
Oh yeah. Getting the event object would be handy. :+1:

**Comment by sukihub** [comment]
(Month and a half long pause because of 1 commit?! Should have finished it sooner.) Anyway, should be ready now :-) https://github.com/cakephp/cakephp/compare/master...sukihub:3.x-conditional-listeners pull request?

**Comment by markstory** [comment]
@sukihub A pull request would be great. Its much easier to review and discuss changes in a pull request.

**Comment by sukihub** [comment]
Here it is #8657

**Comment by markstory** [comment]
Closing as a pull request is up now, and its easier/better to continue the discussion around working code.
